quic/transport/ngtcp2/stream/closedstate.nim

```diff
@@ -1,38 +1,95 @@
 import ../../../basics
 import ../../stream
+import ./helpers
+import ../../framesorter
+import ../native/[connection, errors]
 import chronicles
 
 logScope:
   topics = "closed state"
 
 type
   ClosedStream* = ref object of StreamState
+    stream*: Opt[Stream]
+    remaining: AsyncQueue[seq[byte]]
+    connection*: Ngtcp2Connection
+    frameSorter: FrameSorter
+    cancelRead*: Future[void]
+
   ClosedStreamError* = object of StreamError
 
-proc newClosedStream*(): ClosedStream =
-  ClosedStream()
+proc newClosedStream*(
+    messages: AsyncQueue[seq[byte]],
+    frameSorter: FrameSorter,
+    connection: Ngtcp2Connection,
+): ClosedStream =
+  ClosedStream(
+    remaining: messages,
+    cancelRead: newFuture[void](),
+    frameSorter: frameSorter,
+    connection: connection,
+  )
 
-{.push locks: "unknown".}
+method enter(state: ClosedStream, stream: Stream) =
+  procCall enter(StreamState(state), stream)
+  state.stream = Opt.some(stream)
 
-method enter*(state: ClosedStream, stream: Stream) =
-  procCall StreamState(state).enter(stream)
-  stream.closed.fire()
+proc clearUserData*(state: ClosedStream) =
+  try:
+    setUserData(state.stream, state.connection, nil)
+  except Ngtcp2Error:
+    discard # stream already closed
 
-method read*(state: ClosedStream): Future[seq[byte]] {.async.} =
-  trace "cant read, stream is closed"
-  raise newException(ClosedStreamError, "stream is closed")
+method leave(state: ClosedStream) =
+  state.stream = Opt.none(Stream)
+
+method read(state: ClosedStream): Future[seq[byte]] {.async.} =
+  if not state.frameSorter.isComplete():
+    let incomingFut = state.remaining.get()
+    if (await race(incomingFut, state.cancelRead)) == incomingFut:
+      result = await incomingFut
+      allowMoreIncomingBytes(state.stream, state.connection, result.len.uint64)
+    else:
+      incomingFut.cancelSoon()
+      raise newException(StreamError, "stream is closed")
+  else:
+    try:
+      result = state.remaining.popFirstNoWait()
+      return
+    except AsyncQueueEmptyError:
+      discard
+
+    raise newException(StreamError, "stream is closed")
 
-method write*(state: ClosedStream, bytes: seq[byte]) {.async.} =
+method write(state: ClosedStream, bytes: seq[byte]) {.async.} =
   trace "cant write, stream is closed"
   raise newException(ClosedStreamError, "stream is closed")
 
-method close*(state: ClosedStream) {.async.} =
+method close(state: ClosedStream) {.async.} =
```

Comment:
Why are these no longer public?

Reply:
This I did to make it match OpenState's close function. Even if the function is not public here, it will be still available thru here: https://github.com/vacp2p/nim-quic/blob/main/quic/transport/stream.nim#L30-L31 

Notice that when we use a `Stream`, we call the `close` function of `StreamState` and do not call directly OpenState or ClosedState `close` function, but those will still be called via inheritance